text/3711-required-inlining.md

```diff
@@ -0,0 +1,153 @@
+- Feature Name: `inline(must)` and `inline(required)`
+- Start Date: 2024-09-17
+- RFC PR: [rust-lang/rfcs#3711](https://github.com/rust-lang/rfcs/pull/3711)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+rustc supports inlining functions with the `#[inline]`/`#[inline(always)]`/
+`#[inline(never)]` attributes, but these are only hints to the compiler.
+Sometimes it is desirable to require inlining, not just suggest inlining:
+security APIs can depend on being inlined for their security properties, and
+performance intrinsics can work poorly if not inlined.
+
+# Motivation
+[motivation]: #motivation
+
+rustc's current inlining attributes are only hints, they do not guarantee that
+inlining is performed and users need to check manually whether inlining was
+performed. While this is desirable for the vast majority of use cases, there are
+circumstances where inlining is necessary to maintain the security properties of
+an API, or where performance of intrinsics is likely to suffer if not inlined.
+For example, consider:
+
+- Armv8.3-A's pointer authentication, when used explicitly with intrinsics[^1],
+rely on being inlined to guarantee their security properties. Rust should make
+it easier to guarantee that these intrinsics will be inlined and emit an error
+if that is not possible or has not occurred.
+- Arm's Neon and SVE performance intrinsics have work poorly if not inlined.
+Since users rely on these intrinsics for their application's performance, Rust
+should be able to warn users when these have not been inlined and performance
+will not be as expected.
+
+In general, for many intrinsics, inlining is not an optimisation but an
+important part of the semantics, regardless of optimisation level.
+
+[^1]: Armv8.3-A's pointer authentication can be used implicitly (automatic,
+hint-based) and explicitly (intrinsics). Implicit pointer authentication is
+already implemented by Rust. Explicit pointer authentication intrinsics are not
+yet available in Rust. Exposing pointer authentication intrinsics would require
+this RFC, or something providing equivalent guarantees, as a prerequisite.
```

Comment:
Can you explain or provide a link to *why* these pointer authentication intrinsics require inlining for correctness?

Also my knee-jerk reaction to this need is to... make them actual intrinsics. And not even add the attribute this RFC proposes. If the attribute is going to be perma-unstable/internal-only, what's the difference between that and a lang item?

I do not find this use-case compelling, and I'd much rather the RFC be motivated by something that we can't do using a well-established pattern based on lang items.

Reply:
I don't have a link, but for pointer authentication, the idea would be to upstream some functions into stdarch which use inline assembly[^1] and must be inlined. They have to be inlined because otherwise you're left with a sequence of instructions that can be used as [a gadget](https://llsoftsec.github.io/llsoftsecbook/#return-oriented-programming) to sign whatever you want, so you want to mix those sequences into the functions so they can't be used as a gadget.

For these functions, `inline(required)` would be necessary to make sure that if you're using these intrinsics that they actually have a benefit. The important part of `inline(required)` for this case is really the diagnostic, `inline(always)` gives us about as much inlining. If there's a more limited form of this proposal which - for just these intrinsics - can give us the diagnostic, then that works too.

[^1]: LLVM does have intrinsics that for pointer authentication, but using inline assembly gives us more guarantees around upholding the security properties of the pointer authentication, e.g. calls to identical intrinsics can be optimised away, unencrypted values could be spilled to memory briefly, and using inline assembly can avoid these issues.